### PR TITLE
fix allowed host path in podsecuritypolicy

### LINKED ISSUE #47107: host path volumes allow backsteps, PSP allowedHostPaths only guard prefix
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):BUG REPORT

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.0.184+3837d951919006", GitCommit:"3837d9519190069c44a453c145869e1fced52f23", GitTreeState:"clean", BuildDate:"2017-06-07T06:07:10Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.0.184+3837d951919006", GitCommit:"3837d9519190069c44a453c145869e1fced52f23", GitTreeState:"clean", BuildDate:"2017-06-07T06:07:10Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}

**What happened**:
configure podsecuritypolicy
```
spec:
  allowedHostPaths:
  - /tmp/foo/
```
configure pod
```
  volumes:
      - name: app-logs
        hostPath:
          path: /tmp/foo/../../
```
**What you expected to happen**: 
pod can visit the whole filesystem.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by xilabao**
fixes #47107 
@pweil- @php-coder @liggitt

### CODE REVIEW

File: pkg/security/podsecuritypolicy/util/util.go [hunk 1]
```diff
@@ -182,8 +183,9 @@ func PSPAllowsHostVolumePath(psp *extensions.PodSecurityPolicy, hostPath string)
 		return true
 	}
 
+	path := filepath.Clean(hostPath)
```
Comment (php-coder) [inline pkg/security/podsecuritypolicy/util/util.go]: I remember that `hasPathPrefix()` has some optimizations for not allocated memory and re-use existing slice. I'm wondering if we use `Clean()`  here, will it lower our optimization tactics below?

Another question: would `filepath.Clean()` work for Windows-style paths? (I believe that answer is No.) And should we care about it at all?

Would it be easier to search for `..` and reject such paths without trying to normalize them?

**Comment by php-coder** [comment]
@jhorwit2 JFYI

**Comment by php-coder** [comment]
@xilabao Good catch! Thank you!

File: pkg/security/podsecuritypolicy/util/util.go (hunk 1 above)
Comment (xilabao) [inline pkg/security/podsecuritypolicy/util/util.go]: @php-coder Thanks.
I tried. `filepath.Clean()` can support  Windows-style paths. 
 search for `..` maybe not exact. people may create a dir named `1..2`

Comment (liggitt) [inline pkg/security/podsecuritypolicy/util/util.go]: no cleaning the path... there is no guarantee that the server is running the same OS as the node

Comment (liggitt) [inline pkg/security/podsecuritypolicy/util/util.go]: I'm actually more surprised that hostpath mounts paths with backsteps. @kubernetes/sig-storage-bugs

Comment (smarterclayton) [inline pkg/security/podsecuritypolicy/util/util.go]: Yeah, that seems wrong.  I have trouble imagining how that is allowed, or how it would be helpful.

Comment (smarterclayton) [inline pkg/security/podsecuritypolicy/util/util.go]: I would expect this to be a bug on storage allowing `..` vs something wrong with PSP

Comment (jhorwit2) [inline pkg/security/podsecuritypolicy/util/util.go]: To @liggitt's point, could we do checks and/or cleaning on the path in the kubelet? Similar to how we check for root user in the kubelet?

**Comment by xilabao** [comment]
closed. @jhorwit2 can do better at this issue.

**Pull request closed without merging**